### PR TITLE
Feat/add memory cell

### LINKED ISSUE #109: feat: add `MemoryCell` structure
At the moment, in the `Memory` structure, the `data` field is a vector composed of elements of type `MaybeRelocatable`. 

https://github.com/keep-starknet-strange/cairo-zig/blob/747dc4a492d0f35e4894edb026f6273c7805d0a7/src/vm/memory/memory.zig#L31-L36

In the future, it may be desirable to add other information in addition to only a `MaybeRelocatable` type. In particular, I find the boolean flag `is_accessed` introduced by the Rust implementation quite useful in this context. I therefore propose the implementation of a `MemoryCell` structure which will initially contain the fields:
- `maybeRelocatable` of type `MaybeRelocatable`
- `isAccessed` of type boolean

This structure can then be extended and improved. In this same issue, it will also be necessary to replace the `MaybeRelocatable` type in the `data` vector of the `Memory` structure by the `MemoryCell` structure thus created.

Here are the Rust implementation details [here](https://github.com/lambdaclass/cairo-vm/blob/e6171d66a64146acc16d5512766ae91ae044f297/vm/src/vm/vm_memory/memory.rs#L19-L42).

### PR DESCRIPTION
**Pull request by oxlime**
Closes #109  and Fixes #114

### CODE REVIEW

File: src/vm/core_test.zig [hunk 1]
```diff
@@ -1422,9 +1424,9 @@ test "CairoVM: computeSegmentsEffectiveSizes should return the computed effectiv
     var vm = try CairoVM.init(std.testing.allocator, .{});
     defer vm.deinit();
 
-    try vm.segments.memory.data.put(Relocatable.new(0, 0), .{ .felt = Felt252.fromInteger(1) });
-    try vm.segments.memory.data.put(Relocatable.new(0, 1), .{ .felt = Felt252.fromInteger(1) });
-    try vm.segments.memory.data.put(Relocatable.new(0, 2), .{ .felt = Felt252.fromInteger(1) });
+    try vm.segments.memory.data.put(Relocatable.new(0, 0), MemoryCell.new(.{ .felt = Felt252.fromInteger(1) }));
```
Comment (tcoratger) [inline src/vm/core_test.zig:1427]: To create `MaybeRelocatable` instances in a less verbose way for `Felt252`, you can use the dedicated `fromU256` function that you can find in the `relocatable.zig` file, just to simplify your life.

File: src/vm/memory/memory.zig [hunk 2]
```diff
@@ -16,10 +16,47 @@ const Felt252 = starknet_felt.Felt252;
...
+
+    // Marks Memory Cell as accessed.
+    pub fn markAccessed(self: *Self) void {
+        self.is_accessed = true;
+    }
+
+    // Determines if this Memory Cell is accessed.
+    // # Returns
+    // `true` if it has, `false` otherwise.
+    pub fn isAccessed(self: *const Self) bool {
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:50]: As discussed in #125, I don't think we need this one :)

File: src/vm/memory/memory.zig [hunk 3]
```diff
@@ -16,10 +16,47 @@ const Felt252 = starknet_felt.Felt252;
...
+        return self.is_accessed;
+    }
+
+    // Get MaybeRelocatable value at MemoryCell.
+    pub fn getValue(self: *Self) MaybeRelocatable {
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:55]: As discussed in #125, I don't think we need this one :)

File: src/vm/memory/memory.zig [hunk 4]
```diff
@@ -148,9 +184,20 @@ pub const Memory = struct {
         address: Relocatable,
     ) error{MemoryOutOfBounds}!MaybeRelocatable {
         if (address.segment_index < 0) {
-            return self.temp_data.get(address) orelse CairoVMError.MemoryOutOfBounds;
+            if (self.temp_data.get(address) == null) {
+                return CairoVMError.MemoryOutOfBounds;
+            } else {
+                var tempCell = self.temp_data.get(address).?;
+                return tempCell.getValue();
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:191]: I think that you can directly do:

```zig
return self.temp_data.get(address).?.maybe_relocatable;
```

no?

File: src/vm/memory/memory.zig [hunk 5]
```diff
@@ -148,9 +184,20 @@ pub const Memory = struct {
...
+            }
+        } else {
+            if (self.data.get(address) == null) {
+                return CairoVMError.MemoryOutOfBounds;
+            } else {
+                var cell = self.data.get(address).?;
+                return cell.getValue();
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:198]: I think that you can directly do:

```zig
return self.data.get(address).?.maybe_relocatable;
```
no?

File: src/vm/memory/memory.zig [hunk 6]
```diff
@@ -466,3 +526,35 @@ test "Memory: getRelocatable should return ExpectedRelocatable error if Felt ins
...
+    // Test setup
+    var memory = try Memory.init(std.testing.allocator);
+    defer memory.deinit();
+
+    var cell = MemoryCell.new(MaybeRelocatable{ .relocatable = Relocatable.new(1, 0) });
+
+    // Test checks
+    try expectEqual(
+        cell.getValue(),
+        MaybeRelocatable{ .relocatable = Relocatable.new(1, 0) },
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:540]: In Zig, I know this is confusing, but the order is inversed in the `expectEqual`: expected value is in first position :)

File: src/vm/memory/memory.zig [hunk 7]
```diff
@@ -466,3 +526,35 @@ test "Memory: getRelocatable should return ExpectedRelocatable error if Felt ins
         memory.getRelocatable(Relocatable.new(10, 30)),
     );
 }
+
+test "Memory: Memory Cell getValue" {
+    // Test setup
+    var memory = try Memory.init(std.testing.allocator);
+    defer memory.deinit();
+
+    var cell = MemoryCell.new(MaybeRelocatable{ .relocatable = Relocatable.new(1, 0) });
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:535]: To simplify your life, you can use:

```zig
var cell = MemoryCell.new(fromSegment(1, 0));
```

File: src/vm/core_test.zig (hunk 1 above)
Comment (tcoratger) [reply src/vm/core_test.zig:1427]: Will open an issue to refactor that everywhere for all our unit tests :)

File: src/vm/memory/memory.zig (hunk 4 above)
Comment (oxlime) [reply src/vm/memory/memory.zig:191]: yes much better couldn't call a function but no need for getValue()

File: src/vm/memory/memory.zig (hunk 5 above)
Comment (oxlime) [reply src/vm/memory/memory.zig:198]: as above same

File: src/vm/memory/memory.zig (hunk 7 above)
Comment (oxlime) [reply src/vm/memory/memory.zig]: gotcha

**Review by tcoratger — APPROVED** [review]

**Pull request merged**